### PR TITLE
[BIOMAGE-1925] - Shuffle plot data points

### PR DESCRIPTION
**Pull request by aerlaut**
# Description
**Problem**
Data points in Vega plots are rendered in sequence. This means later cells. This poses a problem because later cells, which maybe in a group, are rendered over the earlier cells, making this later group to look "over-represented" in the plot.

**Solution**
Right now, the categorical and continuous plots generates data by iterating through the groups (e.g. samples/cluster) and appending the cell ids. This causes "ordering" of cell ids by sample/cluster . In code: 
   
    // sample1 = { cellIds: [1, 10]}
    // sample2 = { cellIds: [0, 5 ]}

    const cellIds = []
    samples.forEach((sample) => { cellIds.push(...sample.cellIds) }

    // Result:
    cellIds = [
    1, // from sample 1
    10, // from sample 1
    0, // from sample 2
    5, // from sample 2
    .....
    ]

The changes in this PR uses the order of cells in the embedding data returned from the pipeline. These cells are shuffled. In code:

    // Contains the x,y coordinate of each cell. Null means data is filtered in QC.
    // embeddingData = [[1, 2], null, [5, 6]]
    const data = []
    embeddingData.forEach((coordinates, idx) => { 
        if (!coordinates) return

        data.push({
            cellId: idx,
            coordinates
        }) 
    }

    // Result
    data = [
        {cellId: 0}, // e.g. from sample 2
        {cellId: 2}, // eg.  from sample 1
        ...
    ]

Therefore, the order of rendering follows the order of the shuffled cells generated in the pipeline.

The Embedding in data-exploration is not affected by this issue because it uses the cell ordering as returned in the embeddingData. Therefore, just shuffing the data as done in [pipeline PR #236](https://github.com/hms-dbmi-cellenics/pipeline/pull/236) is enough without further changes in the UI.

# Details
#### URL to issue
https://biomage.atlassian.net/browse/BIOMAGE-1925

#### Link to staging deployment URL (or set N/A)
https://ui-agi-squeaky-termite.scp-staging.biomage.net/

#### Links to any PRs or resources related to this PR
https://github.com/hms-dbmi-cellenics/pipeline/pull/236

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [ ] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [ ] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [ ] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [ ] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [ ] Started end-to-end tests on the latest commit.

### Documentation updates
- [ ] Relevant Github READMEs updated **or** no GitHub README updates required.
- [ ] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

File: src/utils/plotSpecs/generateEmbeddingCategoricalSpec.js [hunk 1]
```diff
@@ -227,7 +227,9 @@ const filterCells = (cellSets, sampleKey, groupBy) => {
   let cellSetLegendsData = [];
   const addedCellSetKeys = new Set();
 
-  filteredCells = filteredCells.map((cell) => {
+  filteredCells = filteredCells.reduce((acc, cell) => {
+    if (!cell) return acc;
```
Comment (cosa65) [inline src/utils/plotSpecs/generateEmbeddingCategoricalSpec.js:231]: What is this guard for? In which cases would we run into a null/undefined cell?

Comment (cosa65) [reply src/utils/plotSpecs/generateEmbeddingCategoricalSpec.js:231]: Oh I see it's to do everything in one iteration instead of a filter after.

Comment (cosa65) [reply src/utils/plotSpecs/generateEmbeddingCategoricalSpec.js:231]: Could you write a small comment with something like "Ignore cells that were filtered"?

File: src/utils/plotSpecs/generateEmbeddingCategoricalSpec.js [hunk 2]
```diff
@@ -227,7 +227,9 @@ const filterCells = (cellSets, sampleKey, groupBy) => {
   let cellSetLegendsData = [];
   const addedCellSetKeys = new Set();
 
-  filteredCells = filteredCells.map((cell) => {
+  filteredCells = filteredCells.reduce((acc, cell) => {
```
Comment (cosa65) [inline src/utils/plotSpecs/generateEmbeddingCategoricalSpec.js:230]: From what I'm getting this wasn't part of the changes, it was a performance improvement, right?

I was thinking, using a `reduce` vs a `map` + `filter` is less declarative. I feel like perhaops we shudl consider using map and filter when this doesn't improve the performance much.

I was reading [this](https://www.davideasaf.com/mapfilter-vs-reduce/) a little bit which I think fits this case.

File: src/utils/plotSpecs/generateEmbeddingCategoricalSpec.js [hunk 3]
```diff
@@ -263,16 +265,16 @@ const filterCells = (cellSets, sampleKey, groupBy) => {
...
-      const { cellId, ...supportingData } = data;
+      const { cellId, coordinates } = data;
 
       return {
-        ...supportingData,
-        x: embeddingData[cellId][0],
-        y: embeddingData[cellId][1],
+        ...filteredCells[cellId],
+        x: coordinates[0],
+        y: coordinates[1],
```
Comment (cosa65) [inline src/utils/plotSpecs/generateEmbeddingCategoricalSpec.js:277]: Nice! way more readable and fixes the issue.
Do you know what was the purpose of `.filter((d) => d.cellId < embeddingData.length)` by the way?

Comment (aerlaut) [reply src/utils/plotSpecs/generateEmbeddingCategoricalSpec.js:277]: I think it was just a short cut to truncate data that's longer than the embedding data returned from the worker.

File: src/utils/plotSpecs/generateEmbeddingCategoricalSpec.js (hunk 2 above)
Comment (aerlaut) [reply src/utils/plotSpecs/generateEmbeddingCategoricalSpec.js:230]: A `map` + `filter` produces arrays, whereas in this case I'm making a dictionary object with the `cellId` as the property, which would make it much easier and faster to access the properties of a cell later on down the line.

This reduce can be replaced with a `forEach` which adds to a resulting object at every iteration. What do you think?

**Comment by aerlaut** [comment]
@alexvpickering the staging link is working, can you please check it out?

File: src/utils/plotSpecs/generateEmbeddingCategoricalSpec.js (hunk 2 above)
Comment (cosa65) [reply src/utils/plotSpecs/generateEmbeddingCategoricalSpec.js:230]: Ah ok, never mind then.

About using a `forEach` or a `reduce` I think both options are fine, if you want to keep it like this I understand now and it makes sense

File: src/utils/plotSpecs/generateEmbeddingCategoricalSpec.js (hunk 3 above)
Comment (cosa65) [reply src/utils/plotSpecs/generateEmbeddingCategoricalSpec.js:277]: yes, although it is still weird, why would we need to perform the truncation

**Review by cosa65 — APPROVED** [review]
lgtm

**Comment by alexvpickering** [comment]
> @alexvpickering the staging link is working, can you please check it out?

Shuffling looks good. Only comment is that a random shuffling (different from what comes from worker) is equally acceptable if performance is an issue for large datasets or it is easier to maintain. I'll leave that to you to choose though :)

**Comment by aerlaut** [comment]
> Shuffling looks good. Only comment is that a random shuffling (different from what comes from worker) is equally acceptable if performance is an issue for large datasets or it is easier to maintain. I'll leave that to you to choose though :)

I think it's better implemented this way as shuffling is done only once in the pipeline and we don't have to do it again in the UI.

**Pull request merged**